### PR TITLE
(PA-6507) Update gem rexml from default to 3.2.9 for CVE-2024-35176

### PR DESCRIPTION
**Pull request by shubhamshinde360**
- The CVE was fixed from rexml version 3.2.7.
 - Patching for the CVE wasn't getting applied cleanly and had a lot of conflicts. So updated the gem version to 3.2.9 in the rexml component file.
 - rexml 3.2.7 requires strscan >= 3.0.9 which contains native extensions. We would need a compiler to build the extensions and there are jruby incompatibilities. This requirement has been relaxed starting from rexml 3.2.9. Therefore we update to rexml 3.2.9 here.
 - Added the change to _shared-agent-components since the CVE impacts both agent-runtime-main (ruby 3.2.4 using rexml 3.2.6) and agent-runtime-7.x (ruby 2.7.8 using rexml 3.2.3)
 - For solaris-10-sparc and solaris-11-sparc, we ignore dependency when installing rexml since the ruby in these platforms tries to install strscan (rexml's dependency) but fails while building native extensions. We can ignore installing strscan since it is shipped with ruby 2.7.8 as its default gem.

### CODE REVIEW

**Comment by shubhamshinde360** [comment]
Kept in draft as testing is pending for impacted projects and platforms.

File: configs/components/rubygem-strscan.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+component 'rubygem-rexml' do |pkg, settings, platform|
+    pkg.version '3.0.9'
+    pkg.md5sum '712528845b4f24bb15989f8964b99cc4'
+    
+    instance_eval File.read('configs/components/_base-rubygem.rb')
+  end
+
```
Comment (bastelfreak) [inline configs/components/rubygem-strscan.rb:7]: can you fix the trailing newlinie?

File: configs/components/rubygem-strscan.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+component 'rubygem-rexml' do |pkg, settings, platform|
```
Comment (cthorn42) [inline configs/components/rubygem-strscan.rb:1]: The component name seems wrong here, shouldn't it be the name of the file?

File: configs/projects/_shared-agent-components.rb [hunk 3]
```diff
@@ -69,3 +69,6 @@
 if platform.is_macos?
   proj.component 'rubygem-CFPropertyList'
 end
+
+proj.component 'rubygem-strscan'
```
Comment (joshcooper) [inline configs/projects/_shared-agent-components.rb:73]: `rexml 3.0.7` does require `strscan >= 3.0.9`, but there was a large backlash from the community, because `strscan 3.x` contains native extensions, which meant you needed a compiler to install `rexml` and there were JRuby incompatibilities.

The `strscan` requirement was later relaxed, so [`rexml` 3.2.9](https://rubygems.org/gems/rexml/versions/3.2.9) now accepts any version of `strscan`.  See https://github.com/ruby/rexml/commit/f1df7d13b3e57a5e059273d2f0870163c08d7420#diff-75d0a867bbfd5e375164c17cf7a9df96761fb050e7308f3918ad6fe1fe12d27d 

In other words, if we bump to `rexml 3.2.9`, then we can drop the `rubygem-strscan` component, since it's a default gem https://stdgems.org/2.7.8/#default-gems.

File: configs/components/rubygem-strscan.rb (hunk 2 above)
Comment (shubhamshinde360) [reply configs/components/rubygem-strscan.rb:1]: Thanks Chris, I missed that.

File: configs/projects/_shared-agent-components.rb (hunk 3 above)
Comment (shubhamshinde360) [reply configs/projects/_shared-agent-components.rb:73]: Thanks @joshcooper,
Yes, I have updated the files in this PR bumping to 3.2.9.
I am currently testing some platforms for sanity with that on my local. I will fully update this PR in a while.

**Comment by shubhamshinde360** [comment]
Ran for impacted projects with all the supported platforms:

Looks like all the platforms pass except solaris platforms with sparc architectures which fails due to strscan dependency not being satisfied.
An example build for this: [solaris-11-sparc, agent-runtime-7.x](https://jenkins-platform.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/3057/BUILD_TARGET=solaris-11-sparc,SLAVE_LABEL=k8s-worker/console)


**agent-runtime-main:**
https://jenkins-platform.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/3056/

Note: I had to abort this build since solaris-11-native-sparc was stuck due to resource allocation. Will re-trigger that laler on. It might fail given solaris sparc arch failures have been observed in 7.x.


**agent-runtime-7.x**
https://jenkins-platform.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/3057/


**pe-bolt-server-runtime-main**
https://jenkins-platform.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/3058/
All build succeeded.

**Review by joshcooper — APPROVED** [review]

**Comment by shubhamshinde360** [comment]
Reran for all impacted projects with their supported platforms:

**[agent-runtime-main](https://jenkins-platform.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/3063/)**


  **[agent-runtime-7.x](https://jenkins-platform.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/3064/)**

**[pe-bolt-server-runtime-main](https://jenkins-platform.delivery.puppetlabs.net/view/vanagon-generic-builder/job/platform_vanagon-generic-builder_vanagon-packaging_generic-builder/3065/)**

Everything finished successfully.

All artifacts can be found at: https://builds.delivery.puppetlabs.net/puppet-runtime/4b0e21d59572b24d85b0ca888a697b08876c0b5a/artifacts/
